cirq-core/cirq/optimizers/measurement_transformers.py

```diff
@@ -0,0 +1,132 @@
+# limitations under the License.
+
+from typing import Any, Dict, List, Set, TYPE_CHECKING, TypeVar, Union
+
+from cirq import circuits, ops, value
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class _MeasurementQid(ops.Qid):
+    """A qubit that substitutes in for a deferred measurement.
+
+    Exactly one qubit will be created per qubit in the measurement gate.
+    """
+
+    def __init__(self, key: Union[str, 'cirq.MeasurementKey'], qid: 'cirq.Qid'):
+        """Initializes the qubit.
+
+        Args:
+            key: The key of the measurement gate being deferred.
+            qid: One qubit that is being measured. Each deferred measurement
+                should create one new _MeasurementQid per qubit being measured
+                by that gate.
+        """
+        self._key = value.MeasurementKey.parse_serialized(key) if isinstance(key, str) else key
+        self._qid = qid
+
+    @property
+    def dimension(self) -> int:
+        return self._qid.dimension
+
+    def _comparison_key(self) -> Any:
+        return (str(self._key), self._qid._comparison_key())
+
+    def __str__(self) -> str:
+        return f"M('{self._key}', q={self._qid})"
+
+    def __repr__(self) -> str:
+        return f'_MeasurementQid({self._key!r}, {self._qid!r})'
+
+
+def defer_measurements(circuit: 'cirq.AbstractCircuit') -> 'cirq.Circuit':
+    """Implements the Deferred Measurement Principle.
+
+    Uses the Deferred Measurement Principle to move all measurements to the
+    end of the circuit.
+    """
+
+    circuit = circuits.CircuitOperation(circuit.freeze()).mapped_circuit(deep=True)
+    qubits_found: Set['cirq.Qid'] = set()
+    terminal_measurements: Set['cirq.MeasurementKey'] = set()
+    control_keys: Set['cirq.MeasurementKey'] = set()
+    for op in reversed(list(circuit.all_operations())):
+        gate = op.gate
+        if isinstance(gate, ops.MeasurementGate):
+            key = value.MeasurementKey.parse_serialized(gate.key)
+            if key not in control_keys and not any(q in qubits_found for q in op.qubits):
+                terminal_measurements.add(key)
+        elif isinstance(op, ops.ClassicallyControlledOperation):
+            for c in op.classical_controls:
+                control_keys.update(c.keys)
+        qubits_found.update(op.qubits)
+    measurement_qubits: Dict['cirq.MeasurementKey', List['_MeasurementQid']] = {}
+
+    def defer(op: 'cirq.Operation') -> 'cirq.OP_TREE':
+        gate = op.gate
+        if isinstance(gate, ops.MeasurementGate):
+            key = value.MeasurementKey.parse_serialized(gate.key)
+            if key in terminal_measurements:
+                return op
+            targets = [_MeasurementQid(key, q) for q in op.qubits]
+            measurement_qubits[key] = targets
+            cxs = [ops.CX(q, target) for q, target in zip(op.qubits, targets)]
+            xs = [ops.X(targets[i]) for i, b in enumerate(gate.invert_mask) if b]  # type: ignore
```

Comment:
Actually, related to the issue with `PauliMeasurementGate` and other arbitrary measurement types, the confusion matrix, and to #4858, I think it could be useful to have `MeasurementGate` decompose into a plain measurement plus the `X` gates that are necessary to represent the invert_mask. Then all we have to do here is decompose the measurement here (for an invert_mask it would decompose to `X(q), measure(q, key), X(q)`) and then the deferral would change it to `X(q), CX(q, target), X(q)`. This allows us to handle any measurement types that can decompose to a standard measurement plus some gates. Kind of kills a bunch of birds with one stone that way.

Reply:
Actually I think it *has* to be done via decomposition: any noisy measurement type could end up leaving the original qubits in some weird state as well, so just adding noise to the ancilla qubits would only be a half solution. (In fact, the invert mask is the one place that only modifying the ancilla qubits works, since invert mask leaves the original qubit unchanged, and I plan on leaving it that way since it's one gate less than decomposing into two X's on the original qubit). I added the decomp implementation for PauliMeasurementGate and it works well.

Reply:
The kind of noise that is being captured by `invert_mask` or `confusion_matrix` is classical readout noise. This means that you actually prepared the qubit in state `1` but ended up reading a `0` (or vice versa) because of error in your readout control system. Note that the state of the underlying qubit after measurement is `1` not `0`. 

Because we are modelling classical noise and not quantum noise, the proposal has been to modify the `MeasurementGate` and take additional parameters (invert_mask / confusion_matrix) which can be used to post-process the classical results obtained after measurement. This is important because we do not want to apply the noise directly to the qubits. 

Therefore, to answer your question, we should not decompose a measurement gate with an invert_mask to X M X because this would mean applying noise to the underlying qubit, which is not what the error mechanism is. 

Note that this kind of "special handling" happens only for measurement gate because that's the only case where we model classical noise. For other types of noise which directly affect the qubits, we use noise models and noise channels to directly apply the noise operations to the qubits -- which is what you are suggesting.

Reply:
Let me make sure what this confusion matrix does. If we have a noisy measurement gate NM, and say it measures everything as `1` for whatever reason, then say we have a circuit `NM(q, key='a'), measure(q, key='b')`. At the end of that will we have `measurements=={'a':1, 'b':0}`? So the noise affects only the cbit, but the qubit behaves as if it was a regular measurement gate?

Reply:
Actually the solution I proposed in 2nd paragraph above doesn't work either, since if we delay the inversion to the end of the circuit, any classical controls based off that inverted measurement would not see it.

Really, I'm not sure if a confusion matrix can be accounted for at all here. We can't delay it for reasons mentioned above. We can't decompose it for reasons you stated. We can't push it into the generated qubits because that would require a complete change all across cirq to have things inspect qubits for this information. We likely just have to reject circuits that have such measurements.

Only other option I can think of is if a confusion matrix can be represented as a channel, and append that channel like we currently do the `X` gates that correspond to the invert_mask. And I think this makes sense--when deferring measurements, the measurements are no longer classical operations. They have to be converted to a quantum equivalent. And I'd expect any noise/inversion/etc on the measurement would have to do the same.

Reply:
Actually that last option seems like it should work, no? Any classical operation should be able to be changed to an equivalent quantum channel in a straightforward manner, I'd expect. If that's the case then I'd think this could be merged as-is, and then creating that quantum channel would be done as part of the confusion matrix feature implementation. Thoughts?